src/EditorFeatures/Core/Implementation/IntelliSense/Completion/AsyncCompletionService.cs

```diff
@@ -93,41 +86,12 @@ public bool TryGetController(ITextView textView, ITextBuffer subjectBuffer, out
             return true;
         }
 
-        private bool UseLegacyCompletion(ITextView textView, ITextBuffer subjectBuffer)
+        private bool UseLegacyCompletion(IFeatureServiceFactory featureServiceFactory, ITextBuffer subjectBuffer)
         {
-            if (!_newCompletionAPIEnabled.HasValue)
-            {
-                int userSetting = 0;
-                const string useAsyncCompletionOptionName = "UseAsyncCompletion";
-                if (textView.Options.GlobalOptions.IsOptionDefined(useAsyncCompletionOptionName, localScopeOnly: false))
-                {
-                    userSetting = textView.Options.GlobalOptions.GetOptionValue<int>(useAsyncCompletionOptionName);
-                }
-
-                // The meaning of the UseAsyncCompletion option definition's values:
-                // -1 - user disabled async completion
-                //  0 - no changes from the user; check the experimentation service for whether to use async completion
-                //  1 - user enabled async completion
-                if (userSetting == 1)
-                {
-                    _newCompletionAPIEnabled = true;
-                }
-                else if (userSetting == -1)
-                {
-                    _newCompletionAPIEnabled = false;
-                }
-                else
-                {
-                    if (Workspace.TryGetWorkspace(subjectBuffer.AsTextContainer(), out var workspace))
-                    {
-                        var experimentationService = workspace.Services.GetService<IExperimentationService>();
-                        _newCompletionAPIEnabled = experimentationService.IsExperimentEnabled(WellKnownExperimentNames.CompletionAPI);
-                    }
-                }
-            }
+            var newCompletionEnabled = featureServiceFactory.GetOrCreate(subjectBuffer).IsEnabled(PredefinedEditorFeatureNames.AsyncCompletion);
```

Comment:
ensure that both setter and getter of the feature state use the same buffer. I wonder if using ITextView might be more reliable

Reply:
@KirillOsenkov do you want to set this property on the `ITextView` or the `ITextBuffer`?

Reply:
ITextView is better I think, but either one works.

Reply:
Done!